### DISCUSSION TITLE
自定义图片缓存规则及匹配查找规则

### DISCUSSION THREAD

**Discussion by 52kg** [onevcat/Kingfisher#2242] [Q&A] [answered]
喵神好，现在我们图片是放到OBS上，OBS有提供了图片瘦身功能，就是在原始URL上拼接相关瘦身参数，请问怎么才能做到，同属原始URL的图片(可能会存在不同大小的多张图片)，始终只保存最大的那张

**Comment by onevcat** [comment]
有更具体一些的例子或者场景么？图片是哪里来的（用户上传？服务预设？）？各种版本图片的URL大概什么样？app 里是如何下载这些图片的,,,之类的信息...

Kingfisher默认使用 URL 的 `absoluteString` 来进行缓存，一般来说从 OBS 获取图片的话会用类似 `?size=axb` 之类的 query 吧。如果没有一些规律，而是完全随机的尺寸的话，可能会比较复杂。

**Reply by 52kg** [reply]
有的，喵神：
比如在列表上显示的是小图(50pt*50pt)，但是在详情里显示的是大图(375pt*375pt)，它们指向的URL是一致的，图片可能有多方上传，最终都是上传到OBS上，然后OBS提供了图片提供了指定大小加载(具体就是在原URL query拼接?x-image-process=image/resize,m_mfit,h_%d,w_%d/imageslim)，这里的h,w对应的是我在具体页面需要用到的大小，然后我现在自定了一个JSBURL实现Resource协议，cacheKey还是url的absoluteString，downloadURL被我改造了一下，如下：
<img width="1403" alt="image" src="https://github.com/onevcat/Kingfisher/assets/57518737/7ba6efd4-87bb-4174-8ac7-25220fbef81e">
调用如下：
<img width="1269" alt="image" src="https://github.com/onevcat/Kingfisher/assets/57518737/9b1d4a32-040f-44a3-803a-5b5a3818fde9">

然后接下来不知道该怎么办了，是要自定义ImageCache，重写里面的retrieveImage吗？喵神

**Reply by onevcat** [reply]
有个疑问...为什么不直接用 `downloadURL` 作为 `cacheKey` 呢？“始终只保存最大的那张” 这个要求是为了什么...磁盘空间么？

**Reply by onevcat** [reply]
（另外，我注意到你用了 `DownsamplingImageProcessor`。其实使用这个的话，最后 cache 的图片就已经是 downsampling 以后的图片了。如果还希望 cache 原图，还需要给一个 `cacheOriginalImage` 的 option。不过这是别的问题了..）

**Reply by 52kg** [reply]
对的 喵神

**Reply by 52kg** [reply]
俺一直以为下采样不会修改缓存数据呢，所以想着那种已缓存大图用在较小视图时下采样处理后显示

**Reply by onevcat** [reply, marked as answer]
> 俺一直以为下采样不会修改缓存数据呢，所以想着那种已缓存大图用在较小视图时下采样处理后显示

这是可以做到的，如果你能确保一开始cache和下载到的就是最大的图，那么只需要 `cacheOriginalImage` 就行了...其他小图的时候，Kingfisher 会智能去找原图，如果有的话，就不会再重新下载了。https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet#using-downsamplingimageprocessor-for-high-resolution-images

不过你原来的这个要求确实相对棘手...需要涉及的状态比较多，（每次尝试缓存前都需要先加载之前的缓存，去比较图片大小）..要么继承 `ImageCache` 去[重载相关方法](https://github.com/onevcat/Kingfisher/blob/master/Sources/Cache/ImageCache.swift#L295)，要么自己额外维护一套“已缓存”的图片尺寸的对应关系...感觉都很棘手。

**Reply by 52kg** [reply]
一开始就保证都下大图，这在现有的设计上感觉太难啦，哈哈哈，喵神，ImageCache retrieveImage函数有办法获取到downloadURL么或者通过options里获取？主要是为了获取需要加载的size

**Reply by 52kg** [reply]
好像也是可以的，俺只压缩图片质量，不去做裁剪，这时候在小视图里下采样处理显示，这也能节省部分存储空间，俺想的第一种方案确实改动太大了，后续如果有好的兼容方案，麻烦喵神di一下哈，感谢喵神解答俺的困惑。
